Dev spi driver

c0nwild opened the pull request:
Closes #280 
This PR introduces basic SPI functionality:
- set chip select pin
- initiate bidirectional data transfer using mosi and miso buffers that have to be provided by the user 

One object per SPI user must be created. CS pin is assigned via the constructor.

Acceptance criteria:

- Unit tests cover 100% and all pass
- Unit tests are complete (no logical fallacy is untested and all equivalency classes were tested)
- SPI class diagram from wiki is not in conflict with implementation
- All automatic checks pass
- Google Styleguide was followed
- Additional guidelines from the wiki were followed
- No logical errors in code
- All error return values are always checked

c0nwild committed:
Merge remote-tracking branch 'origin/dev_spi_interface' into experimental_spi

c0nwild committed:
create test environment for hardware tryouts

c0nwild committed:
Added implementation for nrf24l01 - not compiling atm.

c0nwild committed:
added device specific cpp file to target sources

c0nwild committed:
backmerge

c0nwild committed:
rebasing

Commit:
changed data type for spi transmissions to vector

Commit:
Merge branch 'dev' into experimental_spi
Merge dev into experimental branch for rebasing.

c0nwild committed:
adapted test file

Commit:
changed return behaviour of spi.cpp

c0nwild committed:
miso buffer with fixed length (64 bytes)

c0nwild committed:
deleted unnecessary files

Commit:
use vector as buffer, changed symbol name

Commit:
List of Changes:
- spi.hpp: moved functional definitions from spi_types.h here.
- spi_status.hpp: Corrected status name, added status
- minor refactoring

Commit:
Added doxygen comments, renamed constant

c0nwild committed:
Added private method for setting the CS pin

Commit:
Merge branch 'dev_spi_driver' of https://github.com/squiddrone/squiddrone into dev_spi_driver

Commit:
moved private method declaration out of interface

Commit:
Refactoring spi.hpp/cpp
- moved include directives to header file
- introduced private member for cs pin setting
- added data struct for defining chip select GPIO
- added doxygen documentation

Commit:
Adaptions to spi unit testing
- introduced necessary mocks
- created unit tests

c0nwild committed:
Update main.cpp

Removal of lines for testing purposes

c0nwild committed:
Revert build.sh

Switched document creation back on

Commit:
Changed CSPin definition to typedef

Commit:
Merge branch 'dev_spi_driver' of https://github.com/squiddrone/squiddrone into dev_spi_driver

elheck commented:
Please fix codacy issue

Commit:
Changing typedef for CSPin to anonymous struct

Commit:
changed CSPin struct typedef to alias definition

Commit:
fixed compiler warning

Commit:
Changed CSPin struct alias back to typedef

c0nwild committed:
Rebasing due to changes in Unit testing CMake scripts.

c0nwild committed:
Updated tests/CMakeLists.txt

- created tests for spi
- disabled tests for com buffer and com interface
due to incompatibilty with new spi interface

Commit:
removed files that are not part of this issue

Commit:
Merge branch 'dev_spi_driver' of https://github.com/squiddrone/squiddrone into dev_spi_driver

Commit:
Added unit test for optimal test coverage

ThorsHamster commented on src/spi/spi.cpp:
```diff
@@ -0,0 +1,61 @@
+  if (buffer_size > SPI_TRANSACTION_BUFFER_SIZE_LIMIT) {
+    return_value = true;
+  }
+  return return_value;
```
Could also be

```suggestion
  return buffer_size > SPI_TRANSACTION_BUFFER_SIZE_LIMIT;
```

c0nwild replied:
Refactored.

ThorsHamster commented on src/spi/spi.hpp:
```diff
@@ -0,0 +1,65 @@
+
+namespace spi {
+
+///Maximum buffer size allowed for SPI Transaction
```
Comment is Whitenoise

c0nwild replied:
Do those comments not come up as descriptions in doxygen? At least this was the intention.

c0nwild replied:
Fixed

ThorsHamster commented on src/spi/spi.hpp:
```diff
@@ -0,0 +1,65 @@
+static constexpr std::uint8_t SPI_TRANSACTION_BUFFER_SIZE_LIMIT = 64;
+static constexpr std::uint8_t SPI_HAL_TX_RX_TIMEOUT = 100;
+
+//Valid values for chip select pin setting
```
Comment is Whitenoise

c0nwild replied:
fixed

ThorsHamster commented on src/spi/spi_status.hpp:
```diff
@@ -6,9 +6,11 @@ enum class SPIStatus : int {
   ///SPI transaction was successful.
   SPI_TRANSACTION_SUCCESSFUL,
   ///SPI transaction failed.
```
Comments are whitenoise

ThorsHamster commented on src/types/spi_types.hpp:
```diff
@@ -0,0 +1,10 @@
+
+#include <cinttypes>
+
+namespace types {
```
This is a placeholder for future development?

c0nwild replied:
Hm. On second thought this will hold the SPIStatus enum class instead of spi_status.h.

ThorsHamster commented on tests/CMakeLists.txt:
```diff
@@ -181,8 +195,8 @@ SETUP_ALLTESTS_TARGET(
-        com_message_buffer
+        #com_interface
+        #com_message_buffer
         i2c
```
Why is the package "spi" not listed here?

c0nwild replied:
Added.

ThorsHamster commented on src/spi/spi_interface.hpp:
```diff
@@ -27,11 +28,11 @@ class SPIInterface {
    * @return spi::SPIStatus Status information about the success of the transmission. See \ref SPIStatus for details.
    */
-  virtual auto Transfer(std::unique_ptr<std::uint8_t> RxData, std::unique_ptr<std::uint8_t> TxData) noexcept -> spi::SPIStatus = 0;
+  virtual auto Transfer(std::vector<std::uint8_t> &miso_data_buffer, std::vector<std::uint8_t> &mosi_data_buffer) noexcept -> spi::SPIStatus = 0;
```
In your architecture, the return type is called "DriverStatus"

c0nwild replied:
Remnant from the "common error return value" discussion. Will change it to types::SPIStatus and change architecture accordingly.

ThorsHamster commented on tests/spi/spi_interface_tests.cpp:
```diff
@@ -16,10 +16,10 @@ class SpiInterfaceTests : public ::testing::Test {
 };
 
 TEST_F(SpiInterfaceTests, transfer) {
```
I would like to discuss on monday, if those tests are really necessary. I think they are just overhead. The Interface is also tested with the tests of the "real" concrete implementation.

ThorsHamster commented on src/spi/spi.cpp:
```diff
@@ -0,0 +1,61 @@
+  HAL_StatusTypeDef transmit_receive_ret_value = HAL_ERROR;
+  spi::SPIStatus return_value = spi::SPIStatus::SPI_HAL_ERROR;
+
+  uint8_t transmit_size = static_cast<uint8_t>(mosi_data_buffer.size());
```
What happens if transmit_size is 0. There could be a possible dataloss if transmit_size < miso_data_buffer.size()?!

c0nwild replied:
If transmit_size == 0 -> No SPI activity at all (see HAL function HAL_SPI_TransmitReceive). So no harm there.

ThorsHamster commented on tests/spi/spi_tests.cpp:
```diff
@@ -0,0 +1,43 @@
+};
+
+TEST_F(SPITests, exceeding_buffer_size_limit) {
+  std::vector<uint8_t> too_large_buffer(65);
```
A Test also with a size of 64 and 0 would be helpful. (Edgecases)

c0nwild replied:
I added the testcases as proposed. However, keep in mind that the return value of the Transfer() function is dependent on the return value of the HAL-Mock, which is forced inside the test case. So whenever I tell the mock to signal "OK" and the buffer is not too big, the test case will pass.

ThorsHamster commented on src/spi/spi.hpp:
```diff
@@ -0,0 +1,65 @@
+  /**
+   * @brief Set the Chip select GPIO pin.
+   * 
+   * @param pin_high_or_low The desired pin state. High or low.
```
I'm not sure, if pin_high_or_low is a good name. It reflects its possible states, but a pin can only be high or low. So it is not really necessary. Maybe pin_state or something similar would be better. You also write here, that only High or Low is possible and its an enum class. So only those two values are possible.

c0nwild replied:
pin_state it is

ThorsHamster commented on src/spi/spi.cpp:
```diff
@@ -0,0 +1,61 @@
+auto SPI::SetChipSelectPin(PinSetting pin_high_or_low) noexcept -> void {
+  GPIO_PinState pin_state = GPIO_PIN_RESET;
+
+  switch (pin_high_or_low) {
```
Is a switch case with only two values really suitable?

c0nwild replied:
Refactored.

ThorsHamster commented on src/spi/spi.hpp:
```diff
@@ -0,0 +1,65 @@
+  explicit SPI(const CSPin chip_select) : spi::SPIInterface(), chip_select_(chip_select){};
+  virtual ~SPI() = default;
+
+  auto Transfer(std::vector<uint8_t> &miso_data_buffer, std::vector<uint8_t> &mosi_data_buffer) noexcept -> spi::SPIStatus override;
```
Could you describe the Interface of Transfer here a little bit?
I'm not so familiar with SPI. So as a user of this interface, which buffer have I to use to send and receive data?
How big should they be. You use resize() inside of the method. So is the size of miso_data_buffer somehow relevant?

c0nwild replied:
Transfer() is a pure virtual member of the spi_interface. I will add some more description there.

ThorsHamster reviewed (commented):
I have some questions and suggestions.

Commit:
Changed spi_interface.h:

- include spi_types.hpp instead of spi_status.hpp
- updated doxygen documentation
- adapted namespace for return type

Commit:
Changed spi_types.hpp

- now contains SPIStatus enum class

Commit:
Changed spi.hpp

- removed comments for single members
- added doxygen comments to typedef and enum class
- adapted namespace for member return value
- renamed SetChipSelectPin function parameter

Commit:
Deleted

Commit:
Changed spi.cpp

- adapted namespace of SPIStatus type
- refactored private methods

Commit:
Changes to unit test files:

- adapted unit test files due to implementation changes

Commit:
added spi to alltests target

Commit:
added more test cases around buffer size limit

Commit:
Changes to interface.hpp

-MOSI data buffer can be any length
-MISO data buffer must be at least size of MOSI buffer
-Transfer() return value is now of type DriverStatus

Commit:
Changes to SPI implementation

-adapted Transfer() signature
-new function to check if MISO buffer is smaller than MOSI buffer
-refactored Transfer() implementation

Commit:
removed SPIStatus enum class

Commit:
adapted spi interface and spi tests

Commit:
corrected type in test fixture

ThorsHamster commented on src/spi/spi.hpp:
```diff
@@ -42,19 +38,29 @@ class SPI final : spi::SPIInterface {
 
   /**
-   * @brief Check if buffer size is exceeding maximuzm size.
+   * @brief Check if transcaction length is exceeding maximuzm size.
```
Typo: 
```suggestion
   * @brief Check if transaction length is exceeding maximum size.
```

c0nwild replied:
Got it. Thx!

ThorsHamster reviewed (commented):
Only one typo left. Then we are done.
Good job!

c0nwild committed:
Fixed type in doxygen comment

Co-authored-by: ThorsHamster <48162347+ThorsHamster@users.noreply.github.com>

ThorsHamster reviewed (approved):
Good job!
